Lib/typing.py

```diff
@@ -2000,24 +2000,32 @@ class _ProtocolMeta(ABCMeta):
     def __instancecheck__(cls, instance):
         # We need this method for situations where attributes are
         # assigned in __init__.
+        is_protocol_cls = getattr(cls, "_is_protocol", False)
         if (
-            getattr(cls, '_is_protocol', False) and
+            is_protocol_cls and
             not getattr(cls, '_is_runtime_protocol', False) and
             not _allow_reckless_class_checks(depth=2)
         ):
             raise TypeError("Instance and class checks can only be used with"
                             " @runtime_checkable protocols")
 
-        if ((not getattr(cls, '_is_protocol', False) or
-                _is_callable_members_only(cls)) and
-                issubclass(instance.__class__, cls)):
+        if not is_protocol_cls and issubclass(instance.__class__, cls):
             return True
-        if cls._is_protocol:
+
+        protocol_attrs = _get_protocol_attrs(cls)
```

Comment:
We could possibly cache the result of this call on the protocol class in a `__protocol_attrs__` attribute. But I'd rather look at that in a separate PR so that we can evaluate the impact of it in isolation.

Reply:
> We could possibly cache the result of this call on the protocol class in a `__protocol_attrs__` class attribute. ...

If you decide to tackle this, @beartype provides a [caching mechanism for `__instancecheck__`](https://github.com/beartype/beartype/blob/main/beartype/typing/_typingpep544.py#L129) results that may be worth a look, if only for inspiration. I believe some of the same caveats would apply. (Full disclosure: I am an author. Happy to chat, if helpful.)

Reply:
I share @ilevkivskyi's reservations in https://github.com/python/cpython/issues/74690#issuecomment-1239525629 about caching the whole call to `__instancecheck__`, but I think caching the call to `_get_protocol_attrs` could get us a large speedup without the same risk of changing the behaviour. Anyway, TBD in a future PR :)

Reply:
Is monkey-patching of the protocol class itself something we need to care about here?

Reply:
> Is monkey-patching of the protocol class itself something we need to care about here?

Pretty unlikely imo. But yeah, it's that kind of thing that I'd want to think through, which is why I'd probably like to leave that to another PR :)